File: web/python/examples/painter_order_overlap.py
"""Shared painter-order proof for the Noon browser/Pyodide authoring environment.

Rust counterpart: `cargo run -p noon-native --example painter_order_overlap`.
Both predeclare the same animation and permit deterministic midpoint sampling.
"""

from noon import BLUE, GREEN, RED, Circle, Path, Rectangle, Scene, Vec2, VectorPath, linear

scene = Scene()

# Deliberately cross renderer pipelines at the same canvas location. Semantic
# insertion order requires the filled green vector path to remain above the
# analytic red circle and blue rectangle.
circle = Circle(1.25, color=RED).set_fill(RED, opacity=1.0).set_stroke(None)
rectangle = Rectangle(2.1, 2.1, color=BLUE).set_fill(BLUE, opacity=1.0).set_stroke(None)
scene.add(circle, key="painter.circle")
scene.add(rectangle, key="painter.rectangle")

square = (
    VectorPath()
    .move_to(Vec2(-0.8, -0.8))
    .line_to(Vec2(0.8, -0.8))
    .line_to(Vec2(0.8, 0.8))
    .line_to(Vec2(-0.8, 0.8))
    .close()
)
scene.add(Path(square, fill=GREEN, stroke=GREEN, stroke_width=0.0), key="painter.path")

# Predeclare the shared affine animation so either host can seek deterministically
# without keeping a Python source continuation on the playback path.
target = rectangle.copy().rotate(1.5707963267948966)
animation = scene.declare_live_transform_to(rectangle, target, run_time=1.0, rate_func=linear)
live = scene.live_execution()
end = live.play(animation)
live.advance_to(end)
live.complete()

result = scene

File: crates/noon/src/example_scenes/painter_order_overlap.rs
//! Filled analytic and path geometry share one semantic painter order.
use crate::{
    AnimationOptions, ExecutionSession, RateFunction, Scene, SemanticPaint, SemanticStyle, Vec2,
    VectorPath, BLUE, GREEN, RED,
};

/// Paired with `web/python/examples/painter_order_overlap.py` on native and WASM.
pub fn session() -> Result<ExecutionSession, String> {
    let mut scene = Scene::new();
    let mut circle = scene.circle(1.25)?;
    let mut rectangle = scene.rectangle(2.1, 2.1)?;
    for (object, color) in [(&mut circle, RED), (&mut rectangle, BLUE)] {
        object.set_fill(
            f64::from(color.red),
            f64::from(color.green),
            f64::from(color.blue),
            1.0,
        )?;
        object.disable_stroke()?;
    }
    scene.add(&circle)?;
    scene.add(&rectangle)?;
    let path = scene.path(
        VectorPath::new()
            .move_to(Vec2::new(-0.8, -0.8))
            .line_to(Vec2::new(0.8, -0.8))
            .line_to(Vec2::new(0.8, 0.8))
            .line_to(Vec2::new(-0.8, 0.8))
            .close(),
        SemanticStyle {
            fill: Some(SemanticPaint::Solid(GREEN)),
            fill_opacity: 1.0,
            stroke: None,
            ..SemanticStyle::default()
        },
    )?;
    scene.add(&path)?;
    let mut target = rectangle.target_editor()?;
    target.rotate(std::f64::consts::FRAC_PI_2)?;
    let rotation = scene.declare_transform_to(
        &rectangle,
        &target,
        AnimationOptions::new()
            .run_time(1.0)
            .rate_func(RateFunction::Linear),
    )?;
    let mut session = scene.execution_session().map_err(|e| e.to_string())?;
    let mut live = scene.live(&mut session);
    let segment = live.play_animation(&rotation).map_err(|e| e.to_string())?;
    live.advance_segment_to(segment, segment.end_time())
        .map_err(|e| e.to_string())?;
    live.complete_segment(segment).map_err(|e| e.to_string())?;
    Ok(session)
}

#[cfg(test)]
mod tests {
    #[test]
    fn painter_order_and_rotation_survive_direct_seek() {
        let mut session = super::session().unwrap();
        session.seek(0.5).unwrap();
        let objects = &session.frame().objects;
        assert_eq!(objects.len(), 3);
        assert_eq!(objects[0].style.fill, Some(crate::RED));
        assert_eq!(objects[1].style.fill, Some(crate::BLUE));
        assert_eq!(objects[2].style.fill, Some(crate::GREEN));
        assert!((objects[1].transform.rotation - std::f32::consts::FRAC_PI_4).abs() < 1e-6);
    }
}
